network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -114,367 +114,276 @@ func NewControlMsgValidationInspector(
+		topic := channels.Topic(prune.GetTopicID())
+		if tracker.isDuplicate(topic.String()) {
+			return NewDuplicateFoundErr(fmt.Errorf("duplicate topic found: %s", topic.String()))
+		}
+		tracker.set(topic.String())
+		err = c.validateTopic(from, topic, activeClusterIDS)
 		if err != nil {
-			lg.Error().
-				Err(err).
-				Bool(logging.KeySuspicious, true).
-				Msg("failed to distribute invalid control message notification")
+			return err
 		}
 	}
 	return nil
 }
 
-// getCtrlMsgCount returns the amount of specified control message type in the rpc ControlMessage.
-func (c *ControlMsgValidationInspector) getCtrlMsgCount(ctrlMsgType p2pmsg.ControlMessageType, ctrlMsg *pubsub_pb.ControlMessage) uint64 {
-	switch ctrlMsgType {
-	case p2pmsg.CtrlMsgGraft:
-		return uint64(len(ctrlMsg.GetGraft()))
-	case p2pmsg.CtrlMsgPrune:
-		return uint64(len(ctrlMsg.GetPrune()))
-	case p2pmsg.CtrlMsgIHave:
-		return uint64(len(ctrlMsg.GetIhave()))
-	default:
-		return 0
+// inspectIhave performs topic validation on all ihaves in the control message using the provided validateTopic func while tracking duplicates.
+func (c *ControlMsgValidationInspector) inspectIhave(from peer.ID, ihaves []*pubsub_pb.ControlIHave, activeClusterIDS flow.ChainIDList) error {
+	totalIHaves := len(ihaves)
+	if totalIHaves == 0 {
+		return nil
+	}
+	sampleSize := c.config.IHaveMaxSampleSize
+	if sampleSize > totalIHaves {
+		sampleSize = totalIHaves
 	}
-}
 
-// validateTopics ensures all topics in the specified control message are valid flow topic/channel and no duplicate topics exist.
-// Expected error returns during normal operations:
-//   - channels.InvalidTopicErr: if topic is invalid.
-//   - ErrDuplicateTopic: if a duplicate topic ID is encountered.
-func (c *ControlMsgValidationInspector) validateTopics(from peer.ID, validationConfig *p2pconf.CtrlMsgValidationConfig, ctrlMsg *pubsub_pb.ControlMessage) error {
-	activeClusterIDS := c.tracker.GetActiveClusterIds()
-	switch validationConfig.ControlMsg {
-	case p2pmsg.CtrlMsgGraft:
-		return c.validateGrafts(from, ctrlMsg, activeClusterIDS)
-	case p2pmsg.CtrlMsgPrune:
-		return c.validatePrunes(from, ctrlMsg, activeClusterIDS)
-	case p2pmsg.CtrlMsgIHave:
-		return c.validateIhaves(from, validationConfig, ctrlMsg, activeClusterIDS)
-	default:
-		// sanity check
-		// This should never happen validateTopics is only used to validate GRAFT and PRUNE control message types
-		// if any other control message type is encountered here this indicates invalid state irrecoverable error.
-		c.logger.Fatal().Msg(fmt.Sprintf("encountered invalid control message type in validate topics expected %s, %s or %s got %s", p2pmsg.CtrlMsgGraft, p2pmsg.CtrlMsgPrune, p2pmsg.CtrlMsgIHave, validationConfig.ControlMsg))
+	err := c.performSample(p2pmsg.CtrlMsgIHave, uint(totalIHaves), uint(sampleSize), func(i, j uint) {
+		ihaves[i], ihaves[j] = ihaves[j], ihaves[i]
+	})
+	if err != nil {
+		return fmt.Errorf("failed to sample ihave messages: %w", err)
```

Comment:
Please double-check the returned errors from all `inspect.*` methods. We should only return an error if the inspection fails for a logically expected reason that signals misbehavior. Errors such as those above that are due to unexpected internal operations must not cause a misbehavior notification; rather they should crash the node.

Reply:
https://github.com/onflow/flow-go/pull/4642/commits/2b2080f11da70e550cbc3dec1c60731d31a51a32